### PR TITLE
feat: Added FUNSD dataset

### PR DESCRIPTION
**Pull request by fg-mindee**
Following up on #124, this PR introduces the following modifications:
- added an abstract VisionDataset class
- added the FUNSD dataset inherited from VisionDataset
- added it to the documentation
- added corresponding unittests
- refactored verified file downloading functions to use it both for model and dataset downloading

Any feedback is welcome!

### CODE REVIEW

File: doctr/datasets/core.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        # Download the file if not present
+        archive_path = os.path.join(dataset_cache, file_name)
+
+        if not os.path.exists(archive_path) and not download:
+            raise ValueError("the dataset needs to be downloaded first with download=True")
+
+        archive_path = download_from_url(url, file_name, file_hash, cache_subdir='datasets')
+
+        # Extract to funsd
```
Comment (charlesmindee) [inline doctr/datasets/core.py:37]: why do you mention funsd on this meta-class ?

**Review by charlesmindee — APPROVED** [review]
Thanks for the PR !

File: doctr/datasets/core.py (hunk 1 above)
Comment (fg-mindee) [reply doctr/datasets/core.py]: Good catch, it was a refactoring, I just fixed it!

**Review by charlesmindee — APPROVED** [review]
Thanks!

**Pull request merged**